### PR TITLE
Add terms of service

### PR DESCRIPTION
**Pull request by Gargron**
Allow admins to manage their terms of service in a legally compliant way. Includes a terms of service template that can be used by providing a few necessary details.

Admins can create terms of service from scratch or using our template. Changes to the terms of service go through a versioning system, where "What's changed" can be entered. Users can be notified about the change over e-mail.

### Demo

https://github.com/user-attachments/assets/72223aab-ee73-4c2b-813e-0720f321f703

### E-mails

grafik

___

Fixes MAS-11

### CODE REVIEW

File: config/templates/terms-of-service.md
Comment (ClearlyClaire) [inline config/templates/terms-of-service.md:1]: I wonder if it could make sense making some sections/information optional. Though since the use of the generator is optional and the draft can be freely edited after generation, that's definitely not a big deal.

**Review by ClearlyClaire — COMMENTED** [review]
This looks like a very useful feature and it looks good overall!

I left a few inline comments, though, and I have a few other questions/remarks:
- overall there is almost no test coverage for the added code
- do we want to always show a “Terms of Service” link even if there is no defined Terms of Service? If so, shouldn't we be more explicit about there being no Terms of Service?
- there seem to be no provision for notifying users *ahead* of ToS changes… is that something we would need, or otherwise be useful?

File: app/javascript/mastodon/features/ui/components/link_footer.tsx
Comment (ClearlyClaire) [reply app/javascript/mastodon/features/ui/components/link_footer.tsx]: Can something be added to `/api/v2/instance` (which we already query) to return whether there is a ToS?

**Comment by mjankowski** [comment]
High level thoughts:

- It might help to add this piece by piece instead of all at once, for easier review, and to iterate through one smaller change at a time?
- Could some of this be done in advance of ToS addition? Examples...
	- Removal of invite link from footer
	- Name/key change from i18n and link (or at least rename/drop first)
	- Any mailer style changes not narrowly related only to the new mailer
	- The typescript-conversion portion of link footer change
- This is a large amount of non-trivial new code, and has ~zero spec coverage added here. At absolute minimum, we should cover the "happy path" through all new controllers with system specs and the API addition with request spec ... ideally with more than that at unit level for new worker, policy, helpers, models, etc.

Will leave some smaller style/naming/etc suggestions/questions inline.

File: app/controllers/admin/terms_of_service/drafts_controller.rb
Comment (mjankowski) [inline app/controllers/admin/terms_of_service/drafts_controller.rb:6]: This view is basically a form ... should it be `edit` action?

File: app/controllers/admin/terms_of_service/previews_controller.rb
Comment (mjankowski) [inline app/controllers/admin/terms_of_service/previews_controller.rb:11]: REST/CRUD naming thing ... could be a separate `tests#create` controller action?

File: app/controllers/admin/terms_of_service/previews_controller.rb
Comment (mjankowski) [inline app/controllers/admin/terms_of_service/previews_controller.rb:17]: REST/CRUD naming thing ... could be a separate `distributions#create` (or whatever) controller action?

File: app/controllers/admin/terms_of_service/histories_controller.rb
Comment (mjankowski) [inline app/controllers/admin/terms_of_service/histories_controller.rb:4]: This view is mainly listing these ... could be `index`?

File: app/models/terms_of_service.rb
Comment (mjankowski) [inline app/models/terms_of_service.rb:27]: Naming thing - could better align this by either calling the query method `notification_sent?` or calling the column in db `notified_at` ?

File: app/models/terms_of_service/generator.rb
Comment (mjankowski) [inline app/models/terms_of_service/generator.rb:8]: Alpha-sort this list?

File: app/views/admin/terms_of_service/_links.html.haml
Comment (mjankowski) [inline app/views/admin/terms_of_service/_links.html.haml:4]: Is this icon intentionally the same as the next one, or still WIP?

File: app/views/admin/terms_of_service/drafts/show.html.haml
Comment (mjankowski) [inline app/views/admin/terms_of_service/drafts/show.html.haml:18]: Style only thing, applies to a bunch of spots will leave just once -- can we use the multi-line form for this? (ie, each option on own line) - makes future-diff changes easier to read, keeps us clean on linting line length limits, etc.

File: app/views/admin/terms_of_service/generates/show.html.haml
Comment (mjankowski) [inline app/views/admin/terms_of_service/generates/show.html.haml:19]: In this one and the other form/edit views ... pull out the input fields to a `_form` partial? (and then `render form` from this view)?

File: app/views/admin/terms_of_service/previews/show.html.haml
Comment (mjankowski) [inline app/views/admin/terms_of_service/previews/show.html.haml:19]: I feel like I suggested this elsewhere before and it was not embraced, but thoughts on using "lazy lookup" for these i18n keys? -- ie `t('.send_preview', ...)` here, and corresponding i18n key path?

File: config/routes.rb
Comment (mjankowski) [inline config/routes.rb:237]: I bet these two lines would be so happy to be wrapped in a `with_options to: 'privacy#show'` block!?

File: config/templates/terms-of-service.md
Comment (mjankowski) [inline config/templates/terms-of-service.md:3]: For ease of future-diff-review, readability, and to match the existing privacy policy markdown file ... hard wrap this whole file at 80 chars?

File: spec/models/terms_of_service_spec.rb
Comment (mjankowski) [inline spec/models/terms_of_service_spec.rb:6]: My preference would be to take the first suggestion from this line, but either of them is better than neither of them.

File: config/templates/terms-of-service.md
Comment (pReya) [inline config/templates/terms-of-service.md:3]: The amount of quotes here is not correct: `("Administrator" we," or "us")`

File: app/controllers/admin/terms_of_service/drafts_controller.rb
Comment (Gargron) [reply app/controllers/admin/terms_of_service/drafts_controller.rb:6]: I don't think the URL needs to be `/draft/edit`, which is why I'm going with `show`. There is no practical difference and there are other controllers that use `show` for forms.

File: app/controllers/admin/terms_of_service/histories_controller.rb
Comment (Gargron) [reply app/controllers/admin/terms_of_service/histories_controller.rb:4]: But I don't want the URL to be `/histories`, it makes sense as `/history` singular.

File: app/mailers/user_mailer.rb
Comment (Gargron) [reply app/mailers/user_mailer.rb:212]: I agree it might be a good change, but this can be done in a refactor.

File: app/views/admin/terms_of_service/_links.html.haml
Comment (Gargron) [reply app/views/admin/terms_of_service/_links.html.haml:4]: Intentionally, since they're the same kind of text document.

File: app/views/admin/terms_of_service/drafts/show.html.haml
Comment (Gargron) [reply app/views/admin/terms_of_service/drafts/show.html.haml:18]: My text editor fails to correctly syntax-color HAML with said line breaks, so I avoid that style.

File: app/views/admin/terms_of_service/generates/show.html.haml
Comment (Gargron) [reply app/views/admin/terms_of_service/generates/show.html.haml:19]: This would make sense to me if multiple views reused the form, but this is the only occurence, so I do not really see the benefit.

File: app/views/admin/terms_of_service/previews/show.html.haml
Comment (Gargron) [reply app/views/admin/terms_of_service/previews/show.html.haml]: I don't have a super strong opinion on this but I am more used to writing it out explicitly.

File: config/routes.rb
Comment (Gargron) [reply config/routes.rb:237]: Actually, this benefitted from having a separate controller. Changed.

File: config/templates/terms-of-service.md
Comment (Gargron) [reply config/templates/terms-of-service.md:1]: From my understanding, none of the sections are optional from a lawyer's perspective, so I'm unlikely to get the green light to do that.

File: app/controllers/admin/terms_of_service/histories_controller.rb
Comment (mjankowski) [reply app/controllers/admin/terms_of_service/histories_controller.rb:4]: Fair enough - though the url could be aliased.

Alt idea - `revisions#index`

File: app/views/admin/terms_of_service/generates/show.html.haml
Comment (mjankowski) [reply app/views/admin/terms_of_service/generates/show.html.haml:19]: 👍

It's definitely more clearly useful in a new+edit scenario, but also imho there's a nice mental scanning/parsing benefit when updating the views.

There's also (again IMO) a nice consistency benefit to be had by just doing it everywhere. I think we currently have form partials in most but not quite all spots.

File: app/controllers/admin/terms_of_service/drafts_controller.rb
Comment (Gargron) [reply app/controllers/admin/terms_of_service/drafts_controller.rb:1]: Fair but we don't handle that when editing privacy policy, rules, or other settings. I think it's out of scope.

File: app/javascript/mastodon/features/terms_of_service/index.tsx
Comment (Gargron) [reply app/javascript/mastodon/features/terms_of_service/index.tsx:45]: It should be a rare scenario but probably makes sense to show the 404 page then.

**Review by ClearlyClaire — APPROVED** [review]
Thank you for adding specs, and for making the footer link dependant on there actually being defined terms. This looks good overall.

I don't think you have commented on this concern though:
> There seem to be no provision for notifying users *ahead* of ToS changes… is that something we would need, or otherwise be useful?

**Comment by Gargron** [comment]
> There seem to be no provision for notifying users ahead of ToS changes… is that something we would need, or otherwise be useful?

I believe that this is OK, although an "effective date" could be built on top of this. I'm still waiting on feedback on this PR from our lawyers, although I checked the general flow with them ahead of the implementation. For the sake of minimizing PR maintenance I'd like to merge this and then process any further feedback as follow-up patches.

**Pull request merged**
